### PR TITLE
Cross-bitness in GcInfoEncoder::Build()

### PR DESCRIPTION
**Pull request by echesakovMSFT**
`BitArray` class in `GcInfoEncoder` uses array of `size_t` internally.

In the cross-bitness scenario (#16513) this causes situations when `BitArray::GetHashCode` in cross-bitness mode computes different hash (in comparison with non cross-bitness mode) for semantically equivalent bit-arrays (meaning having same number of bits and the same values). 

It turned out that the order in which `LiveStateHashTable hashMap` is iterated in `GcInfoEncoder::Build` depends  on the `BitArray::GetHashCode` (via `SimplerHashTable::GetIndexForKey`) which causes binary differences in outputs of cross-bitness crossgen and native crossgen.

This PR makes `BitArray::GetHashCode` target-bitness specific.

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@jkotas Can you please review this?

**Comment by jkotas** [comment]
I understand the subtle problem you are trying to fix. But if somebody looks at this, it is pretty non-obvious why this is using target pointer sizes.

I think it either needs a comment describing the problem; or it should be switched to just use int32 all the time. I do not think the JIT actually needs this to be `size_t`.

**Comment by echesakovMSFT** [comment]
@jkotak Can you please tell if I switch to using int32 in `BitArray` all native PE images previously crossgen'd with `BitArray` using `size_t` are still going to be valid (even though they will be binary different, e.g. on x64 due to the aforementioned dependencies in `BitArrayIterator`)?

As far as I understood, we are using RLE in `GcInfoEncoder::Build` and it shouldn't really matter in which order we are iterating `LiveStateHashTable hashMap` but I just want to confirm.

**Comment by jkotas** [comment]
Correct. This should not affect validity of the existing images.

File: src/gcinfo/gcinfoencoder.cpp [hunk 1]
```diff
@@ -111,43 +111,45 @@
 class BitArray
 {
     friend class BitArrayIterator;
+    typedef unsigned __int32 ChunkType;
```
Comment (jkotas) [inline src/gcinfo/gcinfoencoder.cpp]: Nit: You can use C++ standard `uint32_t`. It looks better.

**Comment by echesakovMSFT** [comment]
**OSX10.12 x64 Checked Innerloop Build and Test** (#19001)

**Comment by echesakovMSFT** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**